custom sampler for ConfigurableTask

LSinev opened the pull request:
* Way to control seed of fewshot sampling (may help with https://github.com/EleutherAI/lm-evaluation-harness/issues/1591)
* Added ability for custom sampler for ConfigurableTask

Python tasks can use customized fewshots strategy (slightly modifiable fewshots or any other ideas) by using child classes of `Task` and modification of `fewshot_context`. Migration of such tasks to `ConfigurableTask` with yaml config was not possible due to a lack of fewshot customization. After this PR, a custom fewshot sampler may be set in the config like
```
fewshot_config:
  sampler: !function utils.MyFewshotSampler
```

haileyschoelkopf reviewed (changes requested):
Hi @LSinev , thanks for adding this and sorry it took so long to review!

This looks good, just worried about this breaking users' existing workflows with the list of 3 seeds. If we made this backward compatible for the time being (and state we'll remove the 3-seed option eventually) then LGTM!

LSinev commented on lm_eval/__main__.py:
```diff
@@ -212,14 +212,16 @@ def setup_parser() -> argparse.ArgumentParser:
         "--seed",
-        type=partial(_int_or_none_list_arg_type, 3),
-        default="0,1234,1234",  # for backward compatibility
+        type=partial(_int_or_none_list_arg_type, 4),
```
Ok, I will see how to implement this idea. Thank you.

LSinev replied:
Added backward compatibility with 3-seed setup, and added warning. Will change warning text if needed.

artemorloff commented:
@haileyschoelkopf seems backward compatibility is done!

LSinev commented:
Made it backward compatible. Added seeds report in results file. Also updated code to be compatible with the main branch for ease of merging.
Check this out, please, @haileyschoelkopf

haileyschoelkopf reviewed (approved):
Thank you @LSinev !
